### ISSUE TITLE
How to understand the 'rec' variable in the implementation?

### ISSUE THREAD

**Issue by Hongliaaa** [yining043/VRP-DACT#1]
problem_tsp.py 中的231行： length = (d1 - d2).norm(p=2, dim=2).sum(1)，请问为什么没有计算最后一个节点返回原始点的距离

In line 231 of  problem_tsp.py,  length = (d1 - d2).norm(p=2, dim=2).sum(1)，why the edge from last node to the first node is not calculated?

-------SUMMARY------------
Hi all, I have updated the explanation related to this issue in the [Readme](https://github.com/yining043/VRP-DACT/blob/main/README.md), see 'Documentation'.

We also provide a Jupyter notebook to help you get started and understand our code. Please open the [Jupyter notebook](https://github.com/yining043/VRP-DACT/blob/main/Play_with_DACT.ipynb) for more details.

**Comment by yining043** [comment]
Hi,

感谢对我们工作的关注。
Thanks for your interest in our work.

这里的 d1 和 d2 表示的是所有TSP的边的起点和终点。这里rec的数据格式类似于链表结构，并不是简单的按照每个点在解中的顺序来存储解的。举个例子，rec[3] = 5 表示的是 编号为3的点连接编号为5的点，并不表示 在当前解中的第三个点是5。
Here, d1 and d2 are used to represent the left and right nodes of each TSP edge, respectively. The data format of "rec" here is similar to the linked list structure in Computer Science. Therefore, the order of the nodes in rec does not directly represent the VRP solution. For example, rec[3] = 5 means that the node with index 3 is connected to the node with index 5, and it does not mean that the third node in the current solution is node 5.

你可以用在272行的 get_real_seq(solutions) 拿到当前的解。
kindly refer to line 272 to the function ''get_real_seq'' if you want to decode the current solution back to print.

**Comment by Hongliaaa** [comment]
感谢您的回复。我还是没有太理解您的意思，您之前的工作TSP-improve-master中的get_cost()计算总长度是这样的（length =(d[:, 1:] - d[:, :-1]).norm(p=2, dim=2).sum(1) + (d[:, 0] - d[:, -1]).norm(p=2, dim=1)），这个计算方法是可以理解的，但是DACT中同样的TSP问题求解get_cost()是这样的（length =  (d1  - d2).norm(p=2, dim=2).sum(1)），有点不能理解，希望得到您的回复。感谢！

**Comment by yining043** [comment]
感谢提问。这两个repo的数据结构是不同的。
例如给定一个TSP-6 的 一个解是 0 -> 1 -> 4 -> 2 -> 3 -> 5 -> 0 

在TSP-improve 中，对这个解进行编码，rec = [ 0, 1, 4, 2, 3, 5]，因此需要计算最后一个点回到第一个点的距离。

在VRP-DACT中，对这个解进行编码，rec = [1, 4, 3, 5, 2, 0]，表示 0-1, 1-4, 2-3, 3-5, 4-2, 5-0，这几条边属于当前解。因此计算距离只需要计算(d1 - d2).norm(p=2, dim=2).sum(1)。

**Comment by Hongliaaa** [comment]
您好，VRP-DACT中的解编码方式是将起始点编码在第一个位置和最后一个位置，相当于编码了两个位置，如您给的例子中，解编码 rec=[0, 1, 4, 2, 3, 5, 0]。非常感谢！

**Comment by yining043** [comment]
您好 我不确定您是否理解了我的意思。

在VRP-DACT中， 我并没有将解编码为 起始点占用第一个最后一个位置。在VRP-DACT中我没有用位置来编码，而是用节点的连接关系来编码。您可以再仔细看看我举的例子~

**Comment by Hongliaaa** [comment]
好的，谢谢

**Comment by Hongliaaa** [comment]
我明白了，感谢！

**Comment by Hongliaaa** [comment]
你好，还是路径长度问题，采用同样的数据，
batch= {'coordinates': tensor([[[0.0290, 0.4019],
         [0.2598, 0.3666],
         [0.0583, 0.7006],
         [0.0518, 0.4681],
         [0.6738, 0.3315],
         [0.7837, 0.5631],
         [0.7749, 0.8208],
         [0.2793, 0.6817],
         [0.2837, 0.6567],
         [0.2388, 0.7313],
         [0.6012, 0.3043],
         [0.2548, 0.6294],
         [0.9665, 0.7399],
         [0.4517, 0.4757],
         [0.7842, 0.1525],
         [0.6662, 0.3343],
         [0.7893, 0.3216],
         [0.5247, 0.6688],
         [0.8436, 0.4265],
         [0.9561, 0.0770]]], device='cuda:0')}
solution= tensor([[ 6, 17,  3,  9, 16,  4, 12,  0,  1,  5, 13, 19, 11, 18,  8, 14, 15,  7,
          2, 10]], device='cuda:0')
我利用上述两种方法计算的结果不一样，（VRP-DACT：length1= tensor([8.4153], device='cuda:0')。
TSP-improve：length2= tensor([10.2169], device='cuda:0')），希望得到得到回复，感谢！

**Comment by yining043** [comment]
nihao

**Comment by yining043** [comment]
您好，非常抱歉，您还是没有理解这里的代码逻辑~

对于举例子的这个解，如果需要在VRP-DACT中测长短，需要将其encode为链表的格式，即为
rec = torch.tensor([[ 1,  5, 10,  9, 12, 13, 17,  2, 14, 16, 6, 18,  0, 19, 15,  7,  4,  3,8, 11]])

此时即可得到正确的路径长短：
<img width="706" alt="image" src="https://user-images.githubusercontent.com/64318410/168409319-57c405e4-62b4-40cc-b5fa-7272afa7dbab.png">

若采用TSP-imporve的数据结构，直接赋值rec=solution即可
<img width="675" alt="image" src="https://user-images.githubusercontent.com/64318410/168409375-ea0d8b78-952d-4373-9fa3-e6de42e15399.png">

**Comment by yining043** [comment]
您好，我已经在readme中添加了解释，请看看是否有帮助：
Hi all, I have updated the explanation related to this issue in the Readme.

![image](https://user-images.githubusercontent.com/64318410/168409760-2edb04ef-19ad-4026-abe5-e7d6eb3093dc.png)

**Comment by Hongliaaa** [comment]
您好，非常抱歉打扰一下，我还是有个小问题没有理解。
我的例子中的解为：solution= tensor([[ 6, 17, 3, 9, 16, 4, 12, 0, 1, 5, 13, 19, 11, 18, 8, 14, 15, 7,2, 10]], device='cuda:0')，您说：如果需要在VRP-DACT中测长短，需要将其encode为链表的格式，即为rec = torch.tensor([[ 1, 5, 10, 9, 12, 13, 17, 2, 14, 16, 6, 18, 0, 19, 15, 7, 4, 3,8, 11]])，
我不明白从solution= tensor([[ 6, 17, 3, 9, 16, 4, 12, 0, 1, 5, 13, 19, 11, 18, 8, 14, 15, 7,2, 10]], device='cuda:0') 到
rec = torch.tensor([[ 1, 5, 10, 9, 12, 13, 17, 2, 14, 16, 6, 18, 0, 19, 15, 7, 4, 3,8, 11]])，是怎么得到的？我在您的VRP-DACT代码中验证了一下结果还是不一样，还有一点您代码里就是将 solution（obj = problem.get_costs(batch, solution)）直接传给了rec （def get_costs(self, batch, rec):），这是我的困惑，希望得到回复，感谢！
![image](https://user-images.githubusercontent.com/93971241/168515690-535ab8b8-db63-48f7-a8c5-72ea0bb0e55d.png)

**Comment by yining043** [comment]
您好，在DACT-VRP中，解在初始化时即被编码为了我提到的“链表”结构。同样的，在进行2-opt、swap、insert操作时，也是基于链表结构进行操作的，你可以发现DACT-VRP的2-opt、swap、insert 的代码和TSP-improve中的完全不同。也就是说，在DACT的代码中所有的解都不存在您举例中的solution那样的直接表示方式。采用链表结构的编码，可以使得2-opt能在pytorch中进行并行，相比于TSP-improve中的那种循环有更好的运行速度。希望您可以再仔细体会一下这里的意思。

我再解释一下这里的数据结构。在DACT-VRP，我们用一个数组rec来存储解的方式是按照链表来编码的：给定一个数组rec，如果是TSP-50，那rec数组的长度即为50。其每一个位置 i 存储的 j，代表了编号为 i 的点在当前解中的下一个点是 j 。（注意，在您举例的 solution 数组中，其每一个位置 i 存储的 j，代表了编号为 j 的点在当前解中的绝对位置为 i ， 可以看出两种数据结构是完全不同的。）

从solution= tensor([[ 6, 17, 3, 9, 16, 4, 12, 0, 1, 5, 13, 19, 11, 18, 8, 14, 15, 7,2, 10]], device='cuda:0') 到
rec = torch.tensor([[ 1, 5, 10, 9, 12, 13, 17, 2, 14, 16, 6, 18, 0, 19, 15, 7, 4, 3,8, 11]])，是我手动计算写出来的。在DACT中并不需要这样的转换，因此没有提供函数。正如上文所说，DACT在初始化的时候就按照链表结构来编码和理解每一个产生的rec解。当然，你可以用在272行的 get_real_seq(solutions) 函数将链表表示的解转换回你理解到的solution那样的直接编码。但是请注意，这个转换函数只是为了我们去看解是什么，在DACT的训练和测试中并没有被用到，完全可以删掉这个函数。

希望上述解释能回答您的问题。最重要的核心是：TSP-improve和DACT-VRP的数据结构逻辑不同，您不能套用TSP-improve的设置来尝试理解DACT的代码哦~

**Comment by yining043** [comment]
Hi all,

I have provided a Jupyter notebook to help you get started and understand our code. Please open the [Jupyter notebook](https://github.com/yining043/VRP-DACT/blob/main/Play_with_DACT.ipynb) here. Hope this helps :)
